derive/src/parse.rs

```diff
@@ -458,103 +482,150 @@ fn next_enum(mut source: &mut Peekable<impl Iterator<Item = TokenTree>>) -> Enum
+fn get_bounds(
+    source: &mut Peekable<impl Iterator<Item = TokenTree>>,
+) -> Vec<(String, Vec<String>)> {
+    let mut ret = Vec::new();
 
+    // Angle bracket generics + bounds
     match source.peek() {
         Some(content) if content.to_string() == "<" => {
             source.next();
-            let mut typename = "".to_string(); 
-            let mut generic_bound = String::new();
+            let mut typename: Option<String> = None;
+            let mut boundname: Option<String> = None;
+            let mut generic_bounds = Vec::new();
             let mut in_type = true;
             while let Some(tok) = source.next() {
                 match tok.to_string().as_str() {
                     ">" => {
-                        ret.insert(typename.clone(), generic_bound.clone());
-                        typename.clear();
-                        generic_bound.clear();
-                        break
-                    },
-                    ":" => {
+                        if let Some(boundname) = boundname.take() {
+                            generic_bounds.push(boundname)
+                        }
+                        if let Some(typename) = typename.take() {
+                            ret.push((typename, std::mem::take(&mut generic_bounds)))
+                        }
+                        break;
+                    }
+                    colon @ ":" => {
                         if in_type {
-                            in_type = false
+                            in_type = false;
                         } else {
-                            generic_bound += ":";
+                            if let Some(boundname) = boundname.as_mut() {
+                                *boundname += colon
+                            } else {
+                                boundname = Some(String::from(colon))
+                            }
                         }
-                    },
+                    }
                     "," => {
-                        ret.insert(typename.clone(), generic_bound.clone());
-                        typename.clear();
-                        generic_bound.clear();
+                        if let Some(boundname) = boundname.take() {
+                            generic_bounds.push(boundname)
+                        }
+                        if let Some(typename) = typename.take() {
+                            ret.push((typename, std::mem::take(&mut generic_bounds)))
+                        }
                         in_type = true
-                    },
+                    }
+                    "+" => {
+                        if let Some(boundname) = boundname.take() {
+                            generic_bounds.push(boundname)
+                        }
+                    }
                     c => {
                         if in_type {
-                            typename+=c;
+                            if let Some(typename) = typename.as_mut() {
```

Comment:
I feel like `typename+=c;` + `.clear()`was a bit better? less strings reallocations and a lighter implementaiton code?

Reply:
It was a little clearer, to read, but harder to cover all the edge cases. Using an `Option` with `.take()` elsewhere helped me avoid errors by letting the type system check for me.